Add SEP defining transactions required

leighmcculloch opened the pull request:
### What
Add a SEP that defines the transactions required to form a payment channel, using the new preconditions in CAP-21, and based on the two-way payment channel in CAP-21's rationale section.

### Why
CAP-21 contains a brief description of how to use CAP-21's new preconditions to support a two-way payment channel protocol, but there are more things that participants need to agree on in regards to the transactions and blockchain state. We should document all these things in a new SEP.

We can work on and iterate on the SEP in this repo for the moment until we have prototyped it and have some confidence we want to propose it as a SEP in the stellar-protocol repository.

This SEP contains no discussion of a transport protocol because it is out-of-scope. We aren't at a point where it is really important to document that concretely as any prototypes we develop now will use whatever simple transport mechanism they can to make the prototype work. We'll document that separately in some future milestone as a separate SEP.

Close #5

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,655 @@
+
+- D_i, the _declaration transaction_, declares an intent to execute the
+corresponding closing transaction C_i.  D_i has source account E, sequence
+number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so
```
s_e is a little confusing to me. Say we start at this state:

```
s (E's seqNum) = 1000, i = 0, e = 0
s_i = 1000
s_e = 1000
```

First C_i/D_i get created but not submitted, i =1.
Second C_i/D_i get created and submitted, i = 2, e=1

This second D_i set E.seqNum to 1000 + 2*1 = 1002
This second C_i then sets E.seqNum to 1003

The current state is then:
```
s = 1003, i=2, e=1
s_i = 1004
s_e = 1002
```

So for a third D_i, shouldn't the minSeqNum be 1003, and thus E's current seqNum. And not s_e?

leighmcculloch replied:
We chatted offline about this, so I think this is resolved. The things that we noted are:
 - Once D is submitted, for e.g. D_2 in the example above, the participants don't sign a third transaction set because the channel is closing.
 - s_e doesn't really matter for channel close, so the doc doesn't mention it being updated in the close steps. We could add it for completeness, but I think it's probably unnecessary? If the second D above was submitted, theoretically e would be equal to 2, the iteration number for the transaction set being submitted.

We used the following table to help think about how s_i and s_e change:
|Description                           |Transaction|Tx Seq|E's seq change|E's seq|s   |i change|i  |e change|e  |s_i |s_e |m  |
|--------------------------------------|-----------|------|--------------|-------|----|--------|---|--------|---|----|----|---|
|Setup: 1. I creates escrow account E  |           |      |1000          |1000   |    |        |   |        |   |    |    |2  |
|Setup: 3. Set variable initial states |           |      |              |1000   |1001|0       |0  |0       |0  |1001|1001|2  |
|                                      |           |      |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|Setup: 5. Build formation transaction.|F          |1001  |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|                                      |           |      |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|Setup: 6. Increment i                 |           |      |              |1000   |1001|1       |1  |        |0  |1003|1001|2  |
|Setup: 7. Build and sign C_i          |C_1        |1004  |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|Setup: 8. Build and sign D_i          |D_1        |1003  |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|Setup: 10. Submit F                   |           |      |1001          |1001   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|------                                |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|Update: Increment i                   |           |      |              |1001   |1001|2       |2  |        |0  |1005|1001|2  |
|Update: Build and sign C_i            |C_2        |1006  |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|Update: Build and sign D_i            |D_2        |1005  |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|                                      |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|------                                |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|                                      |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|Update: Increment i                   |           |      |              |1001   |1001|3       |3  |        |0  |1007|1001|2  |
|Update: Build and sign C_i            |C_3        |1008  |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|Update: Build and sign D_i            |D_3        |1007  |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|                                      |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|---                                   |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|                                      |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|Close: Submit D_2                     |           |      |1005          |1005   |1001|        |3  |2       |2  |1007|1005|2  |
|Close: Submit D_3                     |           |      |1007          |1007   |1001|        |3  |3       |3  |1007|1007|2  |
|Close: Submit C_3                     |           |      |1008          |1008   |1001|        |3  |        |3  |1007|1007|2  |
|                                      |           |      |              |1008   |1001|        |3  |        |3  |1007|1007|2  |

Ref: https://docs.google.com/spreadsheets/d/1cOXL3mYdMTNDu0FE8lgNMa0fjJ2T-S2jolLG5Dl0xRM

acharb replied:
yup that answered it thx. Also in case someone else sees this I was wondering about s_e after the close in case _I_ wanted to continue the channel (I was thinking i and e would remain at 3). But after close the variables reset to 0, so s_e after close doesn't matter (as you said)

leighmcculloch replied:
That's a great point. This is an edge case we should make clearer. I added a section about it in ab21927.

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,655 @@
+Created by I.  Jointly controlled by I and R while the channel is open.
+Control is returned to I at close.
+
+- V, the _reserve account_, that holds an amount of native asset contributed
```
a second multisig account eventually owned by R for R's claimable balances makes sense to me

acharb reviewed (commented):
👋  Had 2 Qs. I might be doing something wrong since I've only been able to read through this once or twice, but I think that's the input that you're looking for 😬 

starting i and e at 0 makes sense to me, which the CAP-21 doesn't specify. The use of s_e is where I'm still a little confuzzled

leighmcculloch commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,655 @@
+
+The two participants agree on the following constants:
+
+- m, the _maximum transaction account of an iteration's transaction set_, is
```
Oops, this is a typo. Should be:
```suggestion
- m, the _maximum transaction count for an iteration's transaction set_, is
```
It's the number of transactions that can be in an iteration, which is yup, always 2.

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+
+- C_i, the _closing transaction_, disburses funds to R and changes the signing
+weights on E such that I unilaterally controls E.  C_i has source account E,
+sequence number s_i+1, a `minSeqAge` of O (the observation period time
```
Wouldn't this be `s_i+m`?

leighmcculloch replied:
The C tx needs to be the second tx in the tx set. s_i is the seq of the first tx in the tx set, that's why C is s_i+1.

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+corresponding closing transaction C_i.  D_i has source account E, sequence
+number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so
+long as E's sequence number n satisfies s_e <= n < s_i.  Because C_i has source
+account E and sequence number s_i+1, D_i leaves E in a state where C_i can
```
Same as above: `s_i+m`

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+
+- D_i, see [Update](#Update) process.
+
+#### Update
```
A general note about this section: one of the main reasons for updating a payment channel is to effectively make an off-chain payment between the two participants. 

This is clearly enabled by the mechanics described below (if I correctly understood you simply have to update the value on the `CREATE_CLAIMABLE_BALANCE` operation), but maybe it's worth making this more explicit? After all, I would guess this is the main functionality people are looking for when implementing this protocol.

leighmcculloch replied:
Good call. Added in d653b78.

accordeiro replied:
Looks great!

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+declaration transaction and closing the channel at the actual final state.
+
+1. Get E's sequence number n
+2. If s_{e+1} >= n < s_i, go to step 3, else go to step 1
```
`s_{e+m}`

leighmcculloch replied:
If we wanted to use m here it would be s_e+m, but the intention is to communicate the seq for the next iteration after e, which I think is clearer as it is. If m ever changed too it means the boosting here doesn't have to deal with that.

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+2. If s_{e+1} >= n < s_i, go to step 3, else go to step 1
+3. Submit most recent D_i
+4. Wait observation period O
+5. Submit C_i
```
Maybe it's worth adding a paragraph explaining why this works? Effectively you're submitting a transaction that bumps the sequence number to `s_i` which is higher than `s_{e+n}`, which invalidates the transaction the malicious participant submitted.

leighmcculloch replied:
Great call. I don't say why this works here at all. 24543b0

accordeiro replied:
Perfect, thanks!

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+1. I and R sign and exchange signatures for trustline transaction TA_i.
+2. I or R submit TA_i.
+
+If the remove trustline transaction TA_i fails or is never submitted, there is
```
You probably meant **add** trustline here?

leighmcculloch replied:
Thanks! d4022e5

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+
+1. Increment i.
+2. I and R build the deposit transaction P_i.
+3. Set e' to e.
```
It took me a while to realize `e'` is an auxiliary var you're using, since all the other vars are well-defined concepts here. Maybe change the phrasing to `Store the current value for e on e'`?

leighmcculloch replied:
```suggestion
3. Set e' to the value of e.
```

leighmcculloch replied:
Ah yes this is not particularly clear. Thanks. I used slightly different language in ff8ec83.

accordeiro commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+
+1. Increment i.
+2. I and R build the withdrawal transaction W_i.
+3. Set e' to e.
```
Same comment about `e'` as above.

leighmcculloch replied:
ff8ec83

accordeiro reviewed (commented):
Took a first pass, this is looking pretty good, but we'd probably benefit from having more 👀 on the mechanics. I left a bunch of inline comments.

leighmcculloch commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+3. Set variable initial states:
+   - s to E's sequence number + 1.
+   - i to 0.
+   - e to 0.
```
I don't think so, because it's not a variable, it's a constant with a predefined value.

leighmcculloch commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,631 @@
+
+The two participants agree on the following constants:
+
+- m, the _maximum transaction count for an iteration's transaction set_, is
```
I think the definition of m is worthwhile because otherwise we have magic numbers, in this case 2, throughout the formulas. That was something I found difficult to understand in the original specification of the protocol in CAP-21.

---

I commented throughout about why I don't think it is suitable to use m in the other places.

> we should potentially replace `1` with `m`

Note that m has value 2.

accordeiro replied:
Hm yeah I agree that's better than magic numbers. Makes sense to me!

leighmcculloch commented:
Thanks @acharb @accordeiro. I've addressed all your comments. Please let me know if I didn't address any sufficiently. Could you please also review https://github.com/leighmcculloch/stellar--experimental-payment-channels/pull/1 which is intended to be merged into this PR?

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,657 @@
+every off-chain update of the payment channel state, or on-chain setup, deposit,
+withdrawal, etc.
+
+- e, the _executed iteration number_, is initialized to zero. It is incremented
```
❓ I'm still not wrapping my head around why we need a formal variable e. It only increments when the payment channel is closed, at which point it doesn't matter

And therefore s_e is only either s, or the sequence # that E ends up at at the close

leighmcculloch replied:
e is set to the i of any transaction that the parties agree to submit to the network, so it is used in all the cases where the parties agree to modify the channel after setup, such as withdrawing.

acharb replied:
got it. I'm wondering if we can modify e's definition then, I think that's where my confusion stemmed from. What about

```
e, the executed iteration number, is initially set to zero, and thereafter represents the i of any transaction set the parties agree to submit to the network.
```

because it changes when the agreement happens. Not after an iteration is submitted on-chain (eg. in the deposit by R steps it changes before anything is submitted)

acharb replied:
also, if I'm understanding this right I believe P_i should be bumping E.seqNum to s_e, not s_i. [In which case this line](https://github.com/stellar/experimental-payment-channels/pull/3/files#diff-b01917f0debe4bb9e19259c6e34888dd6c41d4f2da28f49af7fc820ba5a38dc7R478) would be 
```
One BUMP_SEQUENCE operation bumping the sequence number of escrow account E to s_e.
```

leighmcculloch replied:
Ah good point, the description doesn't really say what was intended. Updated in a0b735f. Thank you!

leighmcculloch replied:
> also, if I'm understanding this right I believe P_i should be bumping E.seqNum to s_e, not s_i. [In which case this line](https://github.com/stellar/experimental-payment-channels/pull/3/files#diff-b01917f0debe4bb9e19259c6e34888dd6c41d4f2da28f49af7fc820ba5a38dc7R478) (and the other changing channel setup cases) would be

I think they're the same thing essentially. In the deposit flow in step 2 when P_i is built s_i will be the sequence number of the escrow account we want to bump to. s_e is not currently set to the value we want to bump to, but it will be after in step 4.

acharb replied:
ah good point. When I read the construction of P_i it was after going through the steps. At the point P_i is created in your steps s_i is right 👍

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,657 @@
+- D_i, the _declaration transaction_, declares an intent to execute the
+corresponding closing transaction C_i.  D_i has source account E, sequence
+number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so
+long as E's sequence number n satisfies s_e <= n < s_i.  Because C_i has source
```
❓ going off my s_e question, could't this just be
```suggestion
long as E's sequence number n satisfies s <= n < s_i.  Because C_i has source
```

leighmcculloch replied:
When e is zero, then s_e has the same value as s, but when e is greater than zero, it will have a greater value.

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,657 @@
+7. Sign and exchange a closing transaction C_i, that closes the channel with
+disbursements matching the initial contributions.
+8. Sign and exchange a declaration transaction D_i.
+9. I and R sign and exchange signatures for formation transaction F.
```
💡 this wasn't initially obvious to me, wondering if it' worth adding a note about it
```suggestion
9. I and R sign and exchange signatures for formation transaction F.
    - it's important F is signed/exchanged after C_i and D_i in the case I or R disappear
```

leighmcculloch replied:
Good call, I've added three new sections discussing this at relevant points in the doc. 86a02d0

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,657 @@
+  C_i contains operations:
+  - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
+  account V as a sponsor of future reserves.
+  - One `CREATE_CLAIMABLE_BALANCE` operation for each trustline that is
```
💡 it depends on if we go with the 2 escrow account idea from your [PR](https://github.com/leighmcculloch/stellar--experimental-payment-channels/pull/1), but if not then I think it's worth adding a point as to why this is a claimable balance op and not a payment op (to prevent a failed C_i)

maybe could be added in the `## Security Concerns` section instead

leighmcculloch replied:
Good call, I added a paragraph in the security section where the closing transaction failure is discussed. This really relates to that because the use of a payment could cause the closing transaction to fail. a8ff8e7

acharb commented on specifications/sep-payment-channel-mechanism.md:
```diff
@@ -0,0 +1,657 @@
+number making that transaction invalid.
+
+1. Get E's sequence number n
+2. If s_{e+1} >= n < s_i, go to step 3, else go to step 1
```
would it be easier if we just said `2. If n != s_i go to step 3 because n is at a malicious value`, implying n is not at the value the latest D_i.seqNum?

the else statement of step 2 says go back to step 1 and re-get n. When would it differ?

leighmcculloch replied:
For the majority of the channel lifespan n will not be equal to s_i because the escrow account's sequence number will be far in the past compared to the sequence in the last iteration. That's why we need it to be a range. It's okay if the sequence number isn't s_i, as long as it isn't earlier than s_i _and_ later than the last agreed sequence to execute.

acharb reviewed (commented):
Took another pass after understanding more and lgtm 🙌 

I had 1 question around the need for variable `e` and thus `s_e`. I'm not completely convinced they're needed but I could be missing something. 

If we'd like to name the sequence # that the channel closed at to reference posthumously, couldn't we use something like i_closed? Otherwise e just takes on 2 values 0 (start), and i_closed (channel closed), and I don't think we need a formal variable for i_closed since it's not used during the any of the processes (left the questions inline)

Also I had a couple inline 💡 for things that weren't initially obvious to me, and might be worth adding if it doesn't add tmi

leighmcculloch commented:
@acharb Thanks for the review. I've responded to or addressed your comments. I also applied the same changes to leighmcculloch#1 as well. Please reopen any comment threads if I didn't completely address a concern.

nikhilsaraf reviewed (commented):
LGTM
